### DISCUSSION TITLE
클라이언트에서 JWT 토큰 쿠키에 저장하고 보내기

### DISCUSSION THREAD

**Discussion by Leekyeongmi** [codestates/siphoon-note#135] [Q&A]
__서버와 쿠키를 통해 JWT 토큰을 주고 받는 소통을 하기 위해 다음 3가지 문제를 정의했다.__

1) **서버에서 쿠키를 통해 전달한 accessToken에 접근**해야 한다. (해결불가)
그동안 연습했던 스프린트에서는 res.data를 통해서 accessToken에 접근해왔는데, 쿠키를 통해 전달받기로 한 상황이다. 어떻게 접근할 수 있을까? 글 하단에 어떤 시도를 했는지 코드를 적어보았다. 그 외에 다양한 구글링을 해봤지만, 답을 찾지 못한 상황이다. 

2) **클라이언트에서 서버로부터 받은 accessToken을 쿠키에 저장**한다. (해결가능)

```js
// Root.jsx
import React from 'react';
import App from './App';
import { CookiesProvider } from 'react-cookie';

export default function Root() {
  return (
    <CookiesProvider>
      <App />
    </CookiesProvider>
  );
}
```

```js
// App.jsx
import React from 'react';
import { useCookies } from 'react-cookie';

import NameForm from './NameForm';

function App() {
  const [cookies, setCookie] = useCookies(['name']); // 쿠키에 저장하기

  function onChange(newName) {
    setCookie('name', newName, { path: '/' });
  }

  return (
    <div>
      <NameForm name={cookies.name} onChange={onChange} />
      {cookies.name && <h1>Hello {cookies.name}!</h1>}
    </div>
  );
}

export default App;
```

3) **저장한 accessToken을 불러와서 서버에 전송**한다. (해결가능)
```
get(name, [options])
Get a cookie value // 저장한 쿠키의 accessToken 값을 불러와서 headers에 담아 보낸다.
```
![image](https://user-images.githubusercontent.com/86139013/162426408-26a7a4a1-19c0-417a-bbdf-f7bd7ede71db.png)

-----
문제 1을 해결하기 위해 찾아봤던 스프린트 (section3 token)

**[요청에 응답하는 서버 코드]**
```js
// 서버에서 로그인 요청에 성공하면 refreshToken을 만들어서 쿠키에 담아 보내주고 있는 코드
    res.cookie("refreshToken", refreshToken, {
      sameSite: "none",
      httpOnly: true,
      secure: true,
    });
    res.status(200).json({ data: { accessToken }, message: "ok" });
  }
 ```

**[로그인 성공 후에 NETWORK TAB]**
Set-Cookie에 refreshToken이 담겨있다. 클라이언트 쪽에서 어떻게 접근할 수 있을까?
![image](https://user-images.githubusercontent.com/86139013/162423935-d1fe2006-639c-4877-ae5e-5116438a9ec6.png)

**[로그인 요청하는 클라이언트 코드]**
```js
axios
      .post(
        'https://localhost:4000/login',
        {
          userId: this.state.userId,
          password: this.state.password
        },
        {
          'Content-Type': 'application/json',
          withCredentials: true
        }
      )
      .then((res) => {
        console.log(res);
        this.props.loginHandler(true);
        this.props.issueAccessToken(res.data.data.accessToken);
      })
      .catch((err) => alert(err));
  }
```
서버로부터 성공 응답을 받고 나서 응답 객체를 콘솔에 찍어보았다. cookie에 접근할 수 있는 방법은 모르겠다.
![image](https://user-images.githubusercontent.com/86139013/162424484-d62139c8-2d23-48db-9875-77266e762dd8.png)

**Comment by Leekyeongmi** [comment]
1. access Token을 쿠키에 저장할 때, 만료기한을 설정해서 쿠키에 accessToken이 있다면 로그인 처리가 된 것으로 보기도 하더라. (아마, 이건 로그인 유지 기능인걸까..?) 
2. 클라-서버 응답 요청 사이에 쿠키가 어떤 형태로 주고 받아지는지 여전히 추상적이고 모호하다. 명확하게 알고 있지 못하고 있다..

**Reply by gomarag** [reply]
### Q2. 클라-서버 응답 요청 사이에 쿠키가 어떤 형태로 주고 받아지는지 여전히 추상적이고 모호하다. 명확하게 알고 있지 못하고 있다

이 부분에 대해서 정확하게 알아보려고 **response headers의 Set-Cookie에 담긴 토큰을 클라이언트가 가져올 수 있는 방법**에 대해서 찾아봤어요.

정리해주신 질문 내용을 제가 이해한 바로 고민하시는 부분은,
> 서버가 응답 헤더의 쿠키에 심은 액세스 토큰을 클라이언트가 저장하고 있어야 다른 요청을 보낼 때 요청 헤더에 첨부하여 보낼 수 있다. 
그런데 크롬 개발자 도구의 네트워크 탭에서는 Request Headers의 Set-Cookie에서 refreshToken을 확인할 수 있지만, **서버로부터의 응답 객체를 콘솔에 출력했을 때는 refreshToken을 찾을 수 없었음. 따라서 클라이언트는 서버가 쿠키에 담아서 보낸 토큰을 어떻게 추출할 수 있는지** 고민 중.

이라고 생각했습니다.

지금 저도 찾아본 바
`서버`는 서버가 발급한 토큰을 쿠키에 심을 때 해당 쿠키에 `SameSite=None` 과 `Secure: true` 옵션을 필수로 추가하고
`클라이언트`는 서버에 요청을 보낼 때  `withCredentials: true`을 추가해서 보내면 된다! 입니다.

**클라이언트가 별도로 액세스 토큰을 저장하지 않더라도 브라우저가 알아서 쿠키의 내용을 저장**하고 있다가 `withCredentials: true`를 붙여서 보내면, 보관하던 쿠키의 내용의 알아서 첨부하여 서버에 응답을 보내게 되어있는 것 같습니다. 그래서 쿠키의 유효 만료 시간을 `Max-Age`로 설정해주어야 하는 것 같아요. 이게 가능한 이유는 토큰을 발급한 주체가 서버이고, 클라이언트가 보내는 토큰 역시 같은 서버가 발급한 것이기 때문이라고 하네요.

저희는 지금 cross-domain에서 쿠키를 주고 받기 때문에 클라이언트 측에서 서버가 보내는 secure 쿠키를 읽어들일 수 있는 방법은 없는 것 같아요. 

가장 많이 도움이 됐던 검색 결과에요 https://stackoverflow.com/questions/3342140/cross-domain-cookies/63728389#63728389
2020년에 작성된 내용이라 비교적 최근 내용이라 첨부합니다!

남은 질문은 저도 좀 더 공부해보고 내일 다시 얘기하면 좋을 것 같아요:-)

**Reply by Leekyeongmi** [reply]
아, 쿠키를 요청에 포함하고자 할 때 클라이언트에서는 `withCredentials :true` 옵션을 넣기만 해도 쿠키가 요청 헤더에 포함이 되는군요! 서버의 응답헤더에서도 `Access-Control-Allow-Credentials:true` 설정을 해줘야 하고요! 

정리하면 accessToken을 쿠키로 주고받을 경우, 따로 클라이언트에서 상태로 관리하지 않고, `요청을 보낼 때 쿠키를 포함하겠다는 옵션 추가`만 해주면 되겠군요?

**Reply by gomarag** [reply]
네 맞아요
지금 경미님이 위에 적어주신 토큰 스프린트의 코드에서도 클라이언트의 로그인 요청 코드에 '컨텐츠 타입'과 'withCredentials' 옵션만 들어있는데 Request Header에 쿠키가 알아서 포함되어 있더라고요

**Comment by Leekyeongmi** [comment]
오, 그리고 생각해보니 `오늘의 키워드` 경우에도 서버에서 데이터로 받은 다음에 클라이언트에서 쿠키로 저장하는 방식은 위에 문제 2번의 해결 방법을 사용하면 되겠군요!!! 그럼 체험판 페이지에서도 쿠키에서 불러와서 사용할 수 있을 거 같네요!!!

**Reply by gomarag** [reply]
지금 서버는 로그인한 사용자가 본인의 다이어리 페이지에 접속했을 때(GET '/essays'), todaysWord라는 변수에 '오늘의 키워드'를 유저마다 보내주는데,
그럼 쿠키에 담아서 랜딩페이지에 보내면 될까요?

**Reply by Leekyeongmi** [reply]
흠... 아 그러네요 오늘의 키워드는 로그인한 유저만 받을 수 있도록 설계가 되었었죠? 생각해보니 오늘의 키워드 경우에도 `랭킹 유저 리스트`나 `퍼블릭 에세이 리스트`와 함께 랜딩페이지에서 데이터 객체로 받아도 되었겠네요! 그럼 체험용 페이지에서도 렌더링 할 수 있었을텐데...

그리고 서버에서 쿠키에 담는 경우 클라이언트 쪽에서 접근을 하지 못하니 응답 헤더에 데이터 객체에 보내주셔야 클라에서 접근을 할 수 있을 것 같아요! 

**정리하면, 랜딩 페이지 (GET '/') 요청 시, 데이터 객체에 오늘의 키워드를 보내주시면 체험용/로그인 유저용에서 같은 상태 변수로 동일하게 사용**할 수 있을 듯 하네요!

**Reply by gomarag** [reply]
네 그러면 `GET /` 요청으로 랜딩페이지가 사용할 데이터를 보낼 때, `오늘의 키워드`를 아래와 같은 응답 객체에 담아서 보내겠습니다!

```jsx
{
  "status": 200,
  "message": "Successfully load landing page!",
  "data": {
    "userList": [   ],
    "publicEssayList": [   ],
    📌 "todaysWord: '   '
  }
}
```

**Reply by Leekyeongmi** [reply]
네네! 그럼 클라이언트도 로직 수정해서 반영하겠습니다 :)
